src/hooks/sqlite.ts

```diff
@@ -0,0 +1,125 @@
+function createRecordingProxy<T extends RecordingProxyTarget>(
+  sqliteProxyTarget: T,
+  canHaveRowCallback = false,
+) {
+  return new Proxy(sqliteProxyTarget, {
+    apply(target, thisArg, argArray: Parameters<typeof sqliteProxyTarget>) {
+      // Extract sql. If thisArg is a Statement then it's thisArg.sql.
+      // Otherwise thisArg is a Database and the sql must be the first element of the argArray.
+      let sql;
+      const statement = thisArg as unknown;
+      if (
+        statement &&
+        typeof statement === "object" &&
+        "sql" in statement &&
+        typeof statement.sql === "string"
+      )
+        sql = statement.sql;
+      else {
+        if (argArray.length === 0) throw new TypeError("missing sql query string");
+
+        if (typeof argArray[0] !== "string")
+          throw new TypeError("first argument is expected to be a sql query string");
+
+        sql = argArray[0];
+      }
+      const call = recording.sqlQuery("sqlite", sql);
+      const start = getTime();
+
+      let rowCalbackIndex: number | null = null;
+      let completionCallbackIndex: number | null = null;
+
+      if (
+        canHaveRowCallback &&
+        argArray.length > 1 &&
+        typeof argArray[argArray.length - 2] === "function"
+      )
+        rowCalbackIndex = argArray.length - 2;
+
+      if (argArray.length > 0 && typeof argArray[argArray.length - 1] === "function") {
+        if (canHaveRowCallback && rowCalbackIndex == null) rowCalbackIndex = argArray.length - 1;
+        else completionCallbackIndex = argArray.length - 1;
+      }
+
+      if (canHaveRowCallback && rowCalbackIndex == null) {
+        // We should add a dummy row handler since there isn't already one,
+        // to get second function treated as a completion handler.
+        argArray.push((err: Error | null) => {
+          if (err != null) throw err;
+        });
+      }
+
+      if (completionCallbackIndex != null) {
+        // wrap existing completion callback
+        const originalCompletionCallback: unknown = argArray[completionCallbackIndex];
+        assert(typeof originalCompletionCallback === "function");
+        const callback = new Proxy(originalCompletionCallback, {
+          apply(cTarget, cThisArg, cArgArray): unknown {
+            // first argument is an Error if it's not null
+            if (cArgArray.length > 0 && cArgArray[0] != null) throw cArgArray[0];
+
+            recording.functionReturn(call.id, null, getTime() - start);
+            return Reflect.apply(cTarget, cThisArg, cArgArray);
+          },
+        });
+        argArray[completionCallbackIndex] = callback;
+      } else {
+        // add a completion callback
+        const callback = (err: Error | null) => {
+          if (err) throw err;
+          recording.functionReturn(call.id, null, getTime() - start);
+        };
+        argArray.push(callback);
+      }
+
+      return Reflect.apply(target, thisArg, argArray) as T;
+    },
```

Comment:
This function is quite complicated with the index dance. I would like to see a unit test for it (and use c8 to make sure all the branches are covered) or I'd suggest a different way to structure it to make it easier to follow (pseudo-code-ish, not tested):
```typescript
function popIfFunction(args: unknown[]): Function | undefined {
  if (typeof args[args.length - 1] === 'function') return args.pop();
}

function apply(target, thisArg, argArray) {
  const completionCallback = popIfFunction(argArray);
  const rowCallback = popIfFunction(argArray);
  
  if (needsRowCallback) argArray.push(rowCallback);  // undefined if not specified, should be fine

  const sql = isStatement ? thisArg.sql : argArray[0];

  const call = recording.sqlQuery('sqlite', sql);
  const start = getTime();

  argArray.push((...args) => { recording.functionReturn(call.id, ...); if (completionCallback) return completionCallback(...args); }); // I don't think it makes sense to Proxy here, nobody's going to be inspecting the callback
  
  return Reflect.apply(target, thisArg, argArray);
}
```

Reply:
Here, Statement.each and Database.each methods can have a row callback. They have a completion callback, only when they have a row callback, in this case their last two arguments are functions. So we can't be sure that the last argument is always a completion callback with `const completionCallback = popIfFunction(argArray);`

So the possibilities for `.each` are:
```
// No row callback, no completion callback
.each(...rest);
// Only a row callback
.each(...rest, rc);
// Both a row callback and a completion callback
.each(...rest, rc, cc);
```

I'll try to refactor `createRecordingProxy` to make it more understandable.